### PR TITLE
changefeedccl: add retries to sinkless changefeeds

### LINKED ISSUE #85008: cdc: declarative schema changes break core changefeeds
In 22.2 , we're planning on switching to the declarative schema changer. Some schema changes cause restarts - and will thus cause an error for core changefeeds. 

Jira issue: CRDB-17999

### PR DESCRIPTION
**Pull request by jayshrivastava**
This change updates core/sinkless changefeeds to run in a retry loop, allowing for changefeed restarts in case of transient errors or declarative schema changes. 

See commit notes for more details.

Fixes https://github.com/cockroachdb/cockroach/issues/85008

### CODE REVIEW

**Comment by jayshrivastava** [comment]
Note my 2nd commit will be removed prior to merging. It's just there to force sinkless tests in CI.

**Review by HonoreDB — CHANGES_REQUESTED** [review]
This is beautiful. Only comment is that we need to robustly test my claim that restarting a core changefeed from the schema change boundary that caused the error is always safe (never skips a message). TestChangefeedNemeses (in changefeedccl/nemeses_test.go) does almost that already--I think all you'd need to do is run it on sinkless a few hundred times, and either remove the `disableDeclarativeSchemaChangesForTest` or find some other way of ensuring that schema changes cause a retryable error.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85458#-:-N8_fUj45T6U9GRHThlP:b-1apat)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

File: pkg/ccl/changefeedccl/changefeed_stmt.go [hunk 1]
```diff
@@ -196,10 +202,38 @@ func changefeedPlanHook(
 
 			telemetry.Count(`changefeed.create.core`)
 			logChangefeedCreateTelemetry(ctx, jr)
-			err := distChangefeedFlow(ctx, p, 0 /* jobID */, details, progress, resultsCh)
-			if err != nil {
-				telemetry.Count(`changefeed.core.error`)
+
+			var err error
+			for r := retry.StartWithCtx(ctx, changefeedRetryOptions); r.Next(); {
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/changefeed_stmt.go:207]: Let's move this logic to a separate function.. coreChangefeedWithRetry or some such.

File: pkg/ccl/changefeedccl/changefeed_processors.go [hunk 2]
```diff
@@ -1038,7 +1038,7 @@ func (cf *changeFrontier) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMetad
 				// Detect whether this boundary should be used to kill or restart the
 				// changefeed.
 				if cf.frontier.boundaryType == jobspb.ResolvedSpan_RESTART {
-					err = changefeedbase.MarkRetryableError(err)
+					err = changefeedbase.MarkRetryableErrorWithTimestamp(err, cf.frontier.boundaryTime)
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/changefeed_processors.go:1041]: I think this is great.. and very elegant.. I always have to question myself if we need ts.Next() or not.. I don't think we do.

File: pkg/ccl/changefeedccl/changefeed_test.go [hunk 3]
```diff
@@ -1182,7 +1182,7 @@ func TestChangefeedColumnDropsOnTheSameTableWithMultipleFamilies(t *testing.T) {
 		})
 	}
 
-	cdcTest(t, testFn, feedTestOmitSinks("sinkless"))
+	cdcTest(t, testFn, feedTestForceSink("sinkless"))
```
Comment (miretskiy) [inline pkg/ccl/changefeedccl/changefeed_test.go:1185]: I don't think we need to forceSink.. We want to run this test against all available sinks.

I think the logic is correct; I would recommend running this and other tests under stress... 
```
 ./dev test pkg/ccl/changefeedccl --filter=TestChangefeedBackfillCheckpoint --stress   
```

(Extra bonus if you decide to setup GCE worker, so that you can build run roachprod stress -- that's probably an overkill as long as you run at least few hundred iterations and verify that you've ran sinkless sufficient number of times).

**Review by miretskiy — APPROVED** [review]
This is very nice.

**Review by jayshrivastava — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85458)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava and @miretskiy)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/85458#-N8z-wCbAYONBpAhzp4d:-N8z-wCbAYONBpAhzp4e:bmpvtvd) (no related file):*
@honoredb I ran the nemeses 200 times on my gce worker and it had no failure. This is the script I used:
```
import subprocess

for i in range(200):
        res = subprocess.run('./dev test pkg/ccl/changefeedccl --filter=TestChangefeedNemeses  --stream-output -- --nocache_test_results', shell=True)
        assert res.returncode == 0
```
The latest commit shows the nemeses code I ran, which basically forces sinkless feeds and makes them restart when schema changes happen.

cc @miretskiy I believe this addresses your comment regarding stress testing below because the nemeses test is rigorous. 

___

**Review by jayshrivastava — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85458)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @jayshrivastava, and @miretskiy)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/85458#-N8z-wCbAYONBpAhzp4d:-N8zKokYDKLxxBqkvLw7:baust5p) (no related file):*
<details><summary><i>Previously, jayshrivastava wrote…</i></summary><blockquote>

@honoredb I ran the nemeses 200 times on my gce worker and it had no failure. This is the script I used:
```
import subprocess

for i in range(200):
        res = subprocess.run('./dev test pkg/ccl/changefeedccl --filter=TestChangefeedNemeses  --stream-output -- --nocache_test_results', shell=True)
        assert res.returncode == 0
```
The latest commit shows the nemeses code I ran, which basically forces sinkless feeds and makes them restart when schema changes happen.

cc @miretskiy I believe this addresses your comment regarding stress testing below because the nemeses test is rigorous. 


</blockquote></details>

Update: Going to run it 10k times overnight to verify things work. 

___

**Review by HonoreDB — APPROVED** [review]
Approved (pending those test runs and removing the two temp commits).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85458#-:-N8zMyj77Obf0k1rEzXY:br8ampw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava and @miretskiy)
